apps/drec-api/src/pods/device-group/device-group.service.ts

```diff
@@ -0,0 +1,151 @@
+  private readonly logger = new Logger(DeviceGroupService.name);
+
+  constructor(
+    @InjectRepository(DeviceGroup)
+    private readonly repository: Repository<DeviceGroup>,
+    private deviceService: DeviceService,
+  ) {}
+
+  async getAll(): Promise<DeviceGroup[]> {
+    return this.repository.find();
+  }
+
+  async findById(id: number): Promise<DeviceGroup> {
+    const deviceGroup = await this.findDeviceGroupById(id);
+    deviceGroup.devices = await this.deviceService.findForGroup(deviceGroup.id);
+    return deviceGroup;
+  }
+
+  async findOne(
+    conditions: FindConditions<DeviceGroup>,
+  ): Promise<DeviceGroup | null> {
+    return (await this.repository.findOne(conditions)) ?? null;
+  }
+
+  async create(
+    organizationId: string,
+    data: NewDeviceGroupDTO,
+  ): Promise<DeviceGroup> {
+    await this.checkNameConflict(data.name);
+
+    const group = await this.repository.save({
+      name: data.name,
+      organizationId,
+    });
+
+    // For each device id, add the groupId but make sure they all belong to the same owner
+    const devices = await this.deviceService.findByIds(data.deviceIds);
+    const ownerCode = devices[0].registrant_organisation_code;
+    await Promise.all(
+      devices.map(async (device: Device) => {
+        await this.deviceService.addToGroup(device, group.id, ownerCode);
+      }),
+    );
+
+    return group;
+  }
+
+  async addDevices(id: number, data: DeviceIdsDTO): Promise<DeviceGroup> {
+    const deviceGroup = await this.findDeviceGroupById(id);
+
+    const ownerCode = ((await this.deviceService.findForGroup(id)[0]) as Device)
+      ?.registrant_organisation_code;
+    const devices = await this.deviceService.findByIds(data.deviceIds);
+
+    if (!data?.deviceIds?.length) {
+      return;
+    }
+
+    await Promise.all(
+      devices.map(async (device: Device) => {
+        await this.deviceService.addToGroup(device, id, ownerCode);
+      }),
+    );
+    deviceGroup.devices = await this.deviceService.findForGroup(deviceGroup.id);
+    return deviceGroup;
+  }
+
+  async removeDevices(id: number, data: DeviceIdsDTO): Promise<DeviceGroup> {
+    const deviceGroup = await this.findDeviceGroupById(id);
+
+    if (!data?.deviceIds?.length) {
+      return;
+    }
+
+    await Promise.all(
+      data.deviceIds.map(async (deviceId: number) => {
+        await this.deviceService.removeFromGroup(deviceId, id);
+      }),
+    );
+
+    deviceGroup.devices = await this.deviceService.findForGroup(deviceGroup.id);
+    return deviceGroup;
+  }
+
+  async update(id: number, data: UpdateDeviceGroupDTO): Promise<DeviceGroup> {
+    await this.checkNameConflict(data.name);
+    const deviceGroup = await this.findDeviceGroupById(id);
+
+    deviceGroup.name = data.name;
+    const updatedGroup = await this.repository.save(deviceGroup);
+    updatedGroup.devices = await this.deviceService.findForGroup(
+      deviceGroup.id,
+    );
+    return updatedGroup;
+  }
```

Comment:
Should we check if the user owns the device group that it's trying to update?

Reply:
Should be good now. Changed the getById method to also take into account the organization of the logged in user and added this to each request.